modin/core/storage_formats/pandas/query_compiler_validator.py

```diff
@@ -0,0 +1,131 @@
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+"""
+Module contains ``QueryCompilerTypeCaster`` class.
+
+``QueryCompilerTypeCaster`` is used for va.
+"""
+
+from types import FunctionType, MethodType
+from typing import Any, Dict, Tuple, TypeVar
+
+from modin.core.storage_formats.base.query_compiler import BaseQueryCompiler
+
+Fn = TypeVar("Fn", bound=Any)
+
+
+class QueryCompilerTypeCaster:
+    """
+    Cast all query compiler arguments of a the member function to current query compiler.
+
+    """
+
+    @classmethod
+    def __init_subclass__(
+        cls,
+        **kwargs: Dict,
+    ) -> None:
+        """
+        Apply type casting to all children of ``QueryCompilerTypeCaster``.
+
+        """
+        super().__init_subclass__(**kwargs)
+        apply_argument_casting()(cls)
+
+
+def cast_nested_args_to_current_qc_type(arguments, current_qc):
+    def cast_arg_to_current_qc(arg):
+        current_qc_type = type(current_qc)
+        if isinstance(arg, BaseQueryCompiler) and not isinstance(arg, current_qc_type):
+            data_cls = current_qc._modin_frame
+            return current_qc_type.from_pandas(arg.to_pandas(), data_cls)
+        else:
+            return arg
+
+    if isinstance(arguments, tuple):
+        arguments = list(arguments)
+        arguments = cast_nested_args_to_current_qc_type(arguments, current_qc)
+        return tuple(arguments)
+    if isinstance(arguments, list):
+        for i in range(len(arguments)):
+            if isinstance(arguments[i], (list, dict)):
+                cast_nested_args_to_current_qc_type(arguments[i], current_qc)
+            else:
+                arguments[i] = cast_arg_to_current_qc(arguments[i])
+    elif isinstance(arguments, dict):
+        for key in arguments:
+            if isinstance(arguments[key], (list, dict)):
+                cast_nested_args_to_current_qc_type(arguments[key], current_qc)
+            else:
+                arguments[key] = cast_arg_to_current_qc(arguments[key])
+    return arguments
+
+
+def apply_argument_casting():
+    """
+    Cast args of all functions that are members of query compilers.
+    Returns
+    -------
+    func
+        A decorator function.
+    """
+
+    def decorator(obj: Fn) -> Fn:
+        """Decorate function or class to cast all arguments that are query comilers to the current query compiler"""
+        if isinstance(obj, type):
+            seen: Dict[Any, Any] = {}
```

Comment:
Why do we need this dict? Is there a case when we may not get into `except KeyError` branch?

Reply:
The idea was to handle cases where for same function[ different aliases exists](https://github.com/arunjose696/modin/blob/arun-sqc-interop/modin/core/storage_formats/pandas/native_query_compiler.py#L1045-L1046) in basequerycompiler, This is no longer required and no cases in both the query compilers that would go to except so removing this.